lib/filter/filter-cmp.c

```diff
@@ -25,74 +25,206 @@
 #include "filter/filter-cmp.h"
 #include "filter/filter-expr-grammar.h"
 #include "scratch-buffers.h"
+#include "generic-number.h"
+#include "parse-number.h"
 
 #include <stdlib.h>
 #include <string.h>
-
-#define FCMP_EQ  0x0001
-#define FCMP_LT  0x0002
-#define FCMP_GT  0x0004
-#define FCMP_NUM 0x0010
+#include <math.h>
 
 typedef struct _FilterCmp
 {
   FilterExprNode super;
   LogTemplate *left, *right;
-  gint cmp_op;
+  gint compare_mode;
 } FilterCmp;
 
-
 static gint
-fop_compare(FilterCmp *self, const gchar *left, const gchar *right)
+fop_compare_numeric(const gchar *left, const gchar *right)
 {
-  if (self->cmp_op & FCMP_NUM)
-    {
-      gint l = atoi(left);
-      gint r = atoi(right);
-      if (l == r)
-        return 0;
-      else if (l < r)
-        return -1;
-
-      return 1;
-    }
+  gint l = atoi(left);
+  gint r = atoi(right);
+  if (l == r)
+    return 0;
+  else if (l < r)
+    return -1;
+
+  return 1;
+}
 
+static gint
+fop_compare_string(const gchar *left, const gchar *right)
+{
   return strcmp(left, right);
 }
 
-static gboolean
-fop_cmp_eval(FilterExprNode *s, LogMessage **msgs, gint num_msg, LogTemplateEvalOptions *options)
+static inline gboolean
+_is_object(LogMessageValueType type)
 {
-  FilterCmp *self = (FilterCmp *) s;
-
-  ScratchBuffersMarker marker;
-  GString *left_buf = scratch_buffers_alloc_and_mark(&marker);
-  GString *right_buf = scratch_buffers_alloc();
+  return type == LM_VT_JSON || type == LM_VT_LIST;
+}
 
-  log_template_format_with_context(self->left, msgs, num_msg, options, left_buf);
-  log_template_format_with_context(self->right, msgs, num_msg, options, right_buf);
+static inline gboolean
+_is_string(LogMessageValueType type)
+{
+  return type == LM_VT_STRING;
+}
 
+static gboolean
+_evaluate_comparison(FilterCmp *self, gint cmp)
+{
   gboolean result = FALSE;
 
-  gint cmp = fop_compare(self, left_buf->str, right_buf->str);
-
   if (cmp == 0)
     {
-      result = self->cmp_op & FCMP_EQ;
+      result = self->compare_mode & FCMP_EQ;
     }
   else if (cmp < 0)
     {
-      result = !!(self->cmp_op & FCMP_LT);
+      result = !!(self->compare_mode & FCMP_LT);
     }
   else
     {
-      result = !!(self->cmp_op & FCMP_GT);
+      result = !!(self->compare_mode & FCMP_GT);
+    }
+  return result;
+}
+
+
+/* NOTE: this function mimics JavaScript when converting a value to a
+ * number.  As described here:
+ *
+ *   Primitive types: https://javascript.info/type-conversions
+ *   Objects: https://javascript.info/object-toprimitive
+ */
```

Comment:
If I remember correctly we deviate in some cases from JS. Shouldn't we remove this comment?

Reply:
I think it's worth keeping it and completing the note that the implementation follows JS but with slight modifications.

Reply:
I have improved the comments both here and on top of _evaluate_typed()